### DISCUSSION TITLE
DataLoader generates subgraph contains only labeled nodes?

### DISCUSSION THREAD

**Discussion by JiaruiWang** [pyg-team/pytorch_geometric#4520] [Q&A]
Thank you for this amazing framework and all your hard work on this!

I have questions on the training subgraph from data loaders. 
My task is a node classification problem on a 60M nodes graph. 20M nodes are labeled, and 40M nodes are unlabeled. I created the dataset with a 16M nodes training mask, a 2M nodes validation mask, and a 2M nodes testing mask, out of 20M labeled nodes. 
If I feed the 16M training mask into [NeighborLoader](https://pytorch-geometric.readthedocs.io/en/latest/modules/loader.html#torch_geometric.loader.NeighborLoader) or [GraphSAINTSampler](https://pytorch-geometric.readthedocs.io/en/latest/modules/loader.html#torch_geometric.loader.GraphSAINTSampler), do they generate the subgraph contains only the labeled nodes? or the subgraph is generated from one labeled node and its neighbors which are both labeled and unlabeled Nodes? 
If the subgraph contains only labeled nodes, then the node features and messages from the unlabeled nodes are missing in the training. What's the best practice for this?

Thank you very much

**Comment by Padarn** [comment]
The sampling in `NeighborLoader` will consider the whole graph when building the sample (from the `data` you provide), but the first `batch_size` nodes in the returned batch will be those that you're actually sampling.

For example if you do
```
NeighborLoader(
    data=data,
    input_nodes=mask_index,
    batch_size=5
)
```
Then each batch will have the first 5 nodes from the `mask_index` you provided.

**Reply by JiaruiWang** [reply]
Thank you very much for your answer. 
I understand NeighborLoader now. How about the GraphSAINTSampler? GraphSAINTSampler does not have a parameter input_nodes like NeigborLoader to apply the training mask. I found one answer from rusty1s: https://github.com/pyg-team/pytorch_geometric/discussions/2499#discussioncomment-677944
> You can split your data into inductive training, validation, and test (sub)-graphs via torch_geometric.utils.subgraph directly in the prepare_data method and before initializing the GraphSAINTSampler. However, please note that GraphSAINT can only make use of sampling during training, in particular because nodes may be sampled more than once during a single epoch. For validation and testing, it's therefore best to operate on the complete graph.

Because the GraphSAINTSampler samples a subgraph for each GCN layer within the sampler, should I use the complete graph or training subgraph as input to the GraphSAINTSampler?
Since the training subgraphs might lose the connections to the rest part of the complete graph, will it lose the messages from the completed graph before the GraphSAINTSampler is applied during training?

**Reply by Padarn** [reply]
I think the standard way would be to sample the whole graph, and just mask when you are calculating loss, similar to what is done in this [example](https://github.com/pyg-team/pytorch_geometric/blob/be7b23c17b516c6263fbaa3104e1c289a0a4bfc6/examples/graph_saint.py#L61).
